main.go

```diff
@@ -563,6 +568,18 @@ func (c *controller) populate(hashrings []receive.HashringConfig, statefulsets m
 			var endpoints []string
 
 			for i := 0; i < int(*sts.Spec.Replicas); i++ {
+
+				// Do not add a replica to the hashring if polling fails.
+				if c.options.allowOnlyReadyReplicas {
```

Comment:
I don't think that this section is doing what we think it is doing here.

[c.waitForPod](https://github.com/spaparaju/thanos-receive-controller/blob/b3973a78730ff0dea1c48e699787b4de6e4e20a6/main.go#L545) waits until the pod reaches `Running` [phase](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase), meaning the containers have all started. 

A pod is `Ready` when the application passes the configured `readinessProbe` indicating that it is ready to serve traffic.

Reply:
replicas get to 'Running' status once their readiness probe is successful. Yes, there is [NO pod status](c.options.allowOnlyReadyReplicas) called 'Ready'. Are you referring to change the name of the variable 'c.options.allowOnlyReadyReplicas' to 'c.options.allowOnlyRunningReplicas' ?

Reply:
I think @ianbillett is indeed referring to the difference between pod's [_phase_](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase) and [_conditions_](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions). From my understanding, as @ianbillett said, a pod is considered to be running when all containers are started - but this does not necessarily mean the pod has passed the readiness probe and that it is capable of receiving requests. I'm therefore also wondering if trying to rely on this _condition_ as opposed to _phase_ would make more sense :thinking:.

Reply:
In general, `readiness-probe` is meant for internal, pod life-cycling event triggering  changes to `pod status`. 
If `readiness-probe` is defined, until the `readiness-probe` is evaluated no traffic is sent to the respective pod. Failure of `readiness-probe` results in removing pod IP from all the referencing services so that the [failed pod does not get any traffic](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#:~:text=readinessProbe,the%20initial%20delay%20is%20Failure%20.) and triggers pod status change. 
`readiness-probe` is triggered all through lifetime of a pod , so at any point of time the readiness-probe failures result in pod status change.
I checked the [readiness probe](https://github.com/thanos-io/thanos/blob/fa4a40d5218709dc25c7be86af26100bab0516a6/pkg/receive/handler.go#L190) for `Thanos receive` and its not heavy (like loading cache etc.).